src/cache/inmemory/writeToStore.ts

```diff
@@ -815,6 +815,13 @@ function warnAboutDataLoss(
   invariant.warn(
 `Cache data may be lost when replacing the ${fieldName} field of a ${parentType} object.
 
+This could cause additional (usually avoidable) network requests to fetch data
+that were otherwise cached.
+
+Additionally, if you use watchQuery (React useQuery internally uses watchQuery),
+this might also cause rerendering of frontend components and, in some rare cases, infinite loops.
+https://github.com/apollographql/apollo-client/issues/10992
```

Comment:
```suggestion
This could cause additional (usually avoidable) network requests to fetch data that were otherwise cached.
```

Couple things here:

* Reading this a few times, the 2nd paragraph makes this problem seem very scary. I actually think your first sentence here is good enough, so lets just remove the 2nd paragraph, especially since there is more than just React that uses the underlying core client.
* Since this is in a template string, newlines are honored, so I pushed this sentence to a single line.

Reply:
> the 2nd paragraph makes this problem seem very scary

Hi @jerelmiller , thanks for the suggestion! I think the previous warning message was not "scary" enough to encourage action ;-)

> especially since there is more than just React that uses the underlying core client

This is a good point. But I believe this is potentially a big problem for (especially new) users to `useQuery`. Should we dynamically detect if the environment is React and conditionally show the line?

Additionally, there are some wonderful comments in the original thread https://github.com/apollographql/apollo-client/issues/10992 by @phryneas , for example

> This is something we should probably highlight more in the docs: that id field is by far not unneccessary - it is one of the core features of Apollo Client, and you should add it always, and everywhere, if the entity has one.

And I think my [reproduction on codesandbox ](https://codesandbox.io/s/festive-carlos-m7v5jd) would be helpful for users to better understand the problem as well. Should we keep the link to the GitHub issue?

Reply:
>  Hi @jerelmiller , thanks for the suggestion! I think the previous warning message was not "scary" enough to encourage action ;-)

Tbh., I'd say that *any* 5-line-long very-in-depth error message should encourage enough action. 
We could make it scary, we could add a screen shake and red flashing CSS to the website it comes from, at some point it is the responsibility of the end developer to care and read the linked sources.  
No matter what we do, some people will ignore it. 
I'm a bit afraid of making this message too scary, since there is a balance after which people will get too afraid to touch anything - or start ignoring all other error messages that are similarly important, but less scary.

So in terms of balance, I think "additional network requests" is probably enough, and a developer that would actually see an infinite render loop (and a network request for each render) should be able to connect the dots here, especially as this message gets logged for each of these rerenders, so their console will be full of it.

> Additionally, there are some wonderful comments in the original thread https://github.com/apollographql/apollo-client/issues/10992 by @phryneas , for example

These would probably make more sense in the linked docs pages than in the error message itself.

I am also thinking of adding a test along the lines of "did we already see this __typename in the past, but with an `id` field" and adding a separate error message for that. We cannot detect that an entity type has an `id` field, because we don't know the Schema - but we can remember that in the past an entity had an `id` field and infer that it should always have one.

Reply:
> So in terms of balance, I think "additional network requests" is probably enough, and a developer that would actually see an infinite render loop (and a network request for each render) should be able to connect the dots here

Thanks @phryneas, This makes perfect sense. I think the problem with the original message is, for new users who don't yet fully understand the importance of cache and its role (essentially a messaging backbone, not just a cache), it is not obvious what's the potential consequences of cache lost. By adding "additional network requests", it should be enough to hint the possibility of infinite loop of requests.

> These would probably make more sense in the linked docs pages than in the error message itself.

agree.

> We cannot detect that an entity type has an id field, because we don't know the Schema - but we can remember that in the past an entity had an id field and infer that it should always have one.

this would be great, love to see this happen! I remember @adamesque mentioned cycle detection could help, but it seems your solution would achieve similar result, and probably is a simpler and more elegant way to mitigate the issue.

Reply:
committed the suggestion, let me know if any other changes are prefered before we merge this, thanks!